task/amazon/task.go

```diff
@@ -0,0 +1,336 @@
+	t.Attributes.Status = t.Resources.AutoScalingGroup.Attributes.Status
+	t.Attributes.Events = t.Resources.AutoScalingGroup.Attributes.Events
+	return nil
+}
+
+func (t *Task) Read(ctx context.Context) error {
+	log.Println("[INFO] Reading DefaultVPC...")
+	if err := t.DataSources.DefaultVPC.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading DefaultVPCSubnet...")
+	if err := t.DataSources.DefaultVPCSubnet.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading Image...")
+	if err := t.DataSources.Image.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading Bucket...")
+	if err := t.Resources.Bucket.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading Credentials...")
+	if err := t.DataSources.Credentials.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading SecurityGroup...")
+	if err := t.Resources.SecurityGroup.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading KeyPair...")
+	if err := t.Resources.KeyPair.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading Credentials...")
+	if err := t.DataSources.Credentials.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading LaunchTemplate...")
+	if err := t.Resources.LaunchTemplate.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading AutoScalingGroup...")
+	if err := t.Resources.AutoScalingGroup.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Done!")
+	t.Attributes.Addresses = t.Resources.AutoScalingGroup.Attributes.Addresses
+	t.Attributes.Status = t.Resources.AutoScalingGroup.Attributes.Status
+	t.Attributes.Events = t.Resources.AutoScalingGroup.Attributes.Events
+	return nil
+}
+
+func (t *Task) Delete(ctx context.Context) error {
+	log.Println("[INFO] Downloading Directory...")
+	if t.Attributes.Environment.Directory != "" && t.Read(ctx) == nil {
+		if err := t.Pull(ctx, t.Attributes.Environment.Directory); err != nil && err != universal.NotFoundError {
+			return err
+		}
+	}
+	log.Println("[INFO] Deleting AutoScalingGroup...")
+	if err := t.Resources.AutoScalingGroup.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting LaunchTemplate...")
+	if err := t.Resources.LaunchTemplate.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting KeyPair...")
+	if err := t.Resources.KeyPair.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting SecurityGroup...")
+	if err := t.Resources.SecurityGroup.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting Credentials...")
+	if err := t.DataSources.Credentials.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting Bucket...")
+	if err := t.Resources.Bucket.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting ")
+	return nil
+}
+
+func (t *Task) Logs(ctx context.Context) ([]string, error) {
+	if err := t.Read(ctx); err != nil {
```

Comment:
Below <sup>[citation needed]</sup>

Reply:
```go
remote, err := fs.NewFs(ctx, (*t.DataSources.Credentials.Resource)["RCLONE_REMOTE"])
	if err != nil {
		return nil, err
	}
```

Reply:
Do you mean `err := Read(ctx)` and `if err != nil` in different lines?

Reply:
> _Why not?_

Because it's [a common pattern](https://golang.org/doc/effective_go#if). If we extract `err`–only assignments to a separate line, the next one will have to use a plain `=` assignment and it's going to be... much less consistent.

Reply:
IMHO is better because you get used to the structure and the read is automatically.
We are spreading the code with those shitty !=nil returns that if you follow the format you can automatically read it without stopping

Reply:
This is a ES6–destructuring level debate. 😄

### What you prefer

```go
err := operation()
if err != nil {
  return err
}

err = operation() # note `=` instead of `:=`
if err != nil {
  return err
}
```

### What I prefer

```go
if err := operation(); err != nil {
  return err
}

if err := operation(); err != nil {
  return err
}
```

Keeping variable scope as narrow as possible doesn't look like a bad practice.